### DISCUSSION TITLE
PorousFlow and Units

### DISCUSSION THREAD

**Discussion by Flolaffel** [idaholab/moose#21551] [Q&A Modules: General] [answered]
Hello,

I have a question regarding units in `PorousFlow`.

The unit system I usually use is mm, 1e3 kg, s, N, MPa.

Inside the `FluidProperties`-Module all units have to be SI, so m, kg, s, N, Pa. But `PorousFlowSingleComponentFluid` allows the user to set `pressure_unit = MPa`. Does that mean `FluidProperties` stays in SI and **only** the pressure parameters get multiplied by 10^6? Can I keep my normal unit system with mm and MPa or do I have to look out for anything else? For example is the unit for my permeability still mm^4 afterwards or do I have to change it?

**Comment by WilkAndy** [comment, marked as answer]
This is explained fully at https://mooseframework.inl.gov/source/materials/PorousFlowSingleComponentFluid.html .   I'm glad you're using `PorousFlow`.   What are you modelling?

**Reply by Traiwit** [reply]
just a quick question @WilkAndy 

if I use `time_unit = years`

does this mean my BC [PorousFlowPiecewiseLinearSink] flux_function is in m/years ? 

Thanks!

**Reply by WilkAndy** [reply]
Yep, that's correct @Traiwit

**Reply by Flolaffel** [reply]
@WilkAndy I actually read that page and just wanted to make sure that I got it right, before making some silly mistakes. Did I get it right?

I'm modelling a porous cylinder that has a permeable top surface and gets pushed on by an indenter. So the goal is to restrict fluid outflow in the contact area and make fluid flow free in no-contact zones. It's running but it's a bit unstable sometimes.

**Reply by WilkAndy** [reply]
You're probably doing the right thing, @Flolaffel .   Yes, permeability should be written in mm^2 units.  Remember your boundary tractions and pressures (for the indenter or confining pressures) must be specified in MPa.  Also, applying a force or pressure to an outer boundary is usually less stable than applying a displacement, which might be the reason for your occasional instability.  You could always trial standard units just to check you've done nothing wrong - the most painful aspect will probably be regenerating your mesh on the metre scale instead of mm.

**Reply by WilkAndy** [reply]
Oh, i just noticed you wrote "mm^4" for permeability - not sure if that was a typo, but permeability should be in mm^2.  Also, Newtons = Pa.m^2 = MPa.mm^2, so if you're applying a force with the indenter you won't need to worry about its numerical value changing from standard SI units.

**Reply by Flolaffel** [reply]
Thanks for your help.

My permeability unit wasn't a typo. I guess I just mixed some things up and ended up with a wrong unit.

I don't use any external Force BCs, I use the Contact Module to indent my cylinder. As a proof of concept you could say. I'm not exactly sure what about my simulation introduces the "disturbances" I'm getting but contact could be one of them. Still to be tested.

**Reply by WilkAndy** [reply]
Oh yes, contact is a really difficult thing to simulate.  As an alternative, could you just apply a FunctionDirichletBC to control the displacement of a node or group of nodes, with the `function` approximating the indenter's movement?  At least that'd tell you whether the contact was the chief problem, or something wrong with your cylinder model.

**Reply by Flolaffel** [reply]
Well I tried and: the convergence is problematic but the results seem as expected.

When using contact the convergence is fine actually but the results not so much. There are unexpected jumps in the porepressure. The cylinder is getting compressed all the time and still sometimes porepressure reduces and then jumps back up the next timestep. This doesn't happen with a displacement controlled simulation.

Do you have experience with combining `PorousFlow` and `Contact` in MOOSE? For example is there a way to make the indenter non-porous? Because right now he is porous with 0 permeability.

**Reply by WilkAndy** [reply]
I haven't done anything like you're doing, sorry @Flolaffel .  Yes, you certainly can block-restrict PorousFlow, so its Variables live only on the cylinder, however that means you have to use the `Kernels` and `Materials` with `block = cylinder` because PorousFlow's Actions (eg `PorousFlowFullySaturated` don't have that functionality).  However, a cheap way of doing it is to set porosity = 0 and permeability = 0.  That way (unless you're using unusual Kernels) all the PorousFlow residuals will be zero on the indenter.  That means porepressure will remain fixed in the indenter.  (Porepressure is physically meaningless in the indenter, of course, but you'll still see it there in paraview and MOOSE.)   Currently, with permeability = 0 only, PorousFlow will keep the fluid mass fixed at each indenter node, which may actually be quite numerically difficult if you have realistic fluid properties, since porepressure could rise dramatically as the indenter squashes a bit.

Regarding convergence: have you read https://mooseframework.inl.gov/modules/porous_flow/convergence.html and https://mooseframework.inl.gov/modules/porous_flow/nonlinear_convergence_problems.html ?   I wonder whether you've tried `SimpleFluidProperties` because that almost always removes any problems related to fluid properties (eg, if you're inadvertently going through a phase transition).   I wonder whether you could share your mesh + input file, just for the "no-indenter" (DirichletBC) case?

To me, the porepressure jumping around sounds like the simulation hasn't converged properly, or fluid is flowing to/from the indenter or other boundaries?

a

**Reply by Flolaffel** [reply]
Interesting...in my experience setting permeability = 0 converges while porosity = 0 completely destroys convergence. Something along the lines of `failed factorization of Jacobian`. How could that be? Maybe I'll have to resort to using all Kernels without any actions.

I haven't seen the pages you mentioned yet. But my PorousFlow part is single-phase, fully saturated, `SimpleFluidProperties` without any temperature or phase changes. And I'm using a `PiecewiseLinearSink`. The convergence criteria article is interesing, but I had to set mine very high in order to make the simulation go through. Maybe that is part of the problem. You'll see in my input, I have the absolute and relative criteria at 1e-3. But: the `DirichletBC` input I attached is the one where porepressure actually does what I expect it to, even with mentioned tolerances. In the contact input I had tolerances at 1e-7 and still got the weird behaviour. I don't know whether I can set them even lower than that and still expect convergence. Contact is not exactly easy to get to converge.

```
[Mesh]
  [cube]
    type = GeneratedMeshGenerator
    dim = 2
    ymin = 0
    ymax = 0.5
    nx = 30
    ny = 10
  []
  [sideset_left]
    type = SideSetsFromBoundingBoxGenerator
    input = cube
    bottom_left = '-0.1 0.4 0'
    top_right = '0.5 0.6 0'
    block_id = 0
    boundary_id_old = top
    boundary_id_new = 110
  []
  [sideset_right]
    type = SideSetsFromBoundingBoxGenerator
    input = sideset_left
    bottom_left = '0.5 0.4 0'
    top_right = '1.1 0.6 0'
    block_id = 0
    boundary_id_old = top
    boundary_id_new = 111
  []
  patch_update_strategy = auto
[]

[Problem]
  coord_type = RZ
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
  gravity = '0 0 0'
  PorousFlowDictator = dictator
  volumetric_locking_correction = true
[]

[Variables]
  [porepressure]
  []
[]

[Functions]
  [tdisp_y]
    type = PiecewiseLinear
    x = '0  10   100'
    y = '0 -0.1 -0.1'
  []
[]

[Modules]
  [TensorMechanics/Master]
    [all]
      add_variables = true
      strain = FINITE
    []
  []
  [FluidProperties]
    [simple_fluid]
      type = SimpleFluidProperties
      bulk_modulus = 1e7 #Pa
      density0 = 1000 #rho in kg/m³
      thermal_expansion = 0
      viscosity = 1e-3 #mu in Pa*s
    []
  []
[]

[PorousFlowFullySaturated]
  coupling_type = HydroMechanical
  porepressure = porepressure
  biot_coefficient = 0.8
  displacements = 'disp_x disp_y'
  fp = simple_fluid
  stabilization = Full
  pressure_unit = MPa
[]

[BCs]
  [confinex]
    type = DirichletBC
    variable = disp_x
    value = 0
    boundary = 'left bottom'
  []
  [confiney]
    type = DirichletBC
    variable = disp_y
    value = 0
    boundary = 'bottom'
  []
  [topload]
    type = FunctionDirichletBC
    variable = disp_y
    function = tdisp_y
    boundary = 110
  []    
  [outflow]
    type = PorousFlowPiecewiseLinearSink
    variable = porepressure
    boundary = 111
    pt_vals = '-1e9 1e9'
    multipliers = '-1e9 1e9'
    PT_shift = 0
    flux_function = 10
    fluid_phase = 0
    use_mobility = true
    use_relperm = true
  []
[]

[Materials]
  [elasticity_cube]
    type = ComputeIsotropicElasticityTensor
    shear_modulus = 0.5
    bulk_modulus = 320
  []
  [stress]
    type = ComputeFiniteStrainElasticStress
  []
  [porosity_cube]
    type = PorousFlowPorosity
    fluid = true
    mechanical = true
    porosity_zero = 0.5 
    biot_coefficient = 0.8 
    solid_bulk = 320
  []
  [permeability_cube]
    type = PorousFlowPermeabilityConst
    permeability = '0.25e-9 0 0   0 0.25e-9 0   0 0 0.25e-9'
  []
[]

[Preconditioning]
  [andy]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = 'lu       mumps'
    
    #petsc_options_iname = '-pc_type -pc_hypre_type'
    #petsc_options_value = 'hypre boomeramg'
    
    #petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -sub_pc_factor_shift_type -pc_asm_overlap'
    #petsc_options_value = 'gmres      asm      lu           NONZERO                   2             '
  []
[]

[Executioner]
  type = Transient
  solve_type = Newton
  start_time = 0
  end_time = 10
  dt = 0.5  
  nl_max_its = 100
  nl_abs_tol = 1e-3
  nl_rel_tol = 1e-3
[]

[Outputs]
  csv = true
  exodus = true
[]
```

Last but not least: Fluid should not be able to reach the indenter since it has no permeability and I restrict the outflow where contact is happening with a Sink (the process of restricting outflow on the left half of the top boundary while using contact differs from the input I attached here). Maybe there is a problem with the sink in the contact simulation. But the effect only happens with "high" bulk moduli for the porous media. Using up to a few MPa is no Problem but at around 10 MPa it starts acting weird. You can see I used over 300 MPa with the Dirichlet-Input and it worked fine.

**Reply by WilkAndy** [reply]
It's the weekend here, so i won't reply any further except to say.... oh crap, you're right about the porosity=0=permeability thing.  The residual is zero, but so is the Jacobian, so you're essentially solving 0 * delta_p = 0.   If you like, you could add

```
[Kernels]
  [dummy_p]
    type = TimeDerivative
    variable = porepressure
    block = indenter
  []
[]
```

Or something similar, just so your jacobian isn't zero.

**Reply by WilkAndy** [reply]
Hi @Flolaffel ,

Below is an alternate input file.  You can compare mine with yours to see the differences:

- do not use the TensorMechanics Action, because PorousFlowFullySaturated already adds Kernels and stuff related to mechanics.  HOWEVER, my version uses small strain, which i know you don't want eventually.  If you want to large large strain, you can't use PorousFlowFullySaturated
- use of small strain and `volumetric_locking_correction = false` improves convergence dramatically
- please consider your numerical quantities, most specifically the `flux_function` in the `PorousFlowPiecewiseLinearSink`.  This is a **major** contributor to non-convergence.  It's impossible for me to know what `flux_function` is appropriate for you.  You might find the `PorousFlowOutflow` is actually a better BC for your case, especially if you include a small layer of very high perm elements to simulate the boundary to air.

After addressing the `PorousFlowPiecewiseLinearSink` issue using my version, I suggest looking at my version, and slowly changing it into your version (eg, using Kernels and finite strain) to compare the convergence.

Finally, well done for building your input file :-)  The changes i mention are small compared with all the work you've put into this already.

```
[Mesh]
  [cube]
    type = GeneratedMeshGenerator
    dim = 2
    ymin = 0
    ymax = 0.5
    nx = 30
    ny = 10
  []
  [sideset_left]
    type = SideSetsFromBoundingBoxGenerator
    input = cube
    bottom_left = '-0.1 0.4 0'
    top_right = '0.5 0.6 0'
    block_id = 0
    boundary_id_old = top
    boundary_id_new = 110
  []
  [sideset_right]
    type = SideSetsFromBoundingBoxGenerator
    input = sideset_left
    bottom_left = '0.5 0.4 0'
    top_right = '1.1 0.6 0'
    block_id = 0
    boundary_id_old = top
    boundary_id_new = 111
  []
  patch_update_strategy = auto
[]

[Problem]
  coord_type = RZ
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
  gravity = '0 0 0'
  PorousFlowDictator = dictator
# NOTE  volumetric_locking_correction = true
[]

[Variables]
  [porepressure]
  []
# NOTE following additions due to not using TensorMechanics Action
  [disp_x]
  []
  [disp_y]
  []
[]

[Functions]
  [tdisp_y]
    type = PiecewiseLinear
    x = '0  10   100'
    y = '0 -0.1 -0.1'
  []
[]

[Modules]
# NOTE: do not need tensor mech action in my case because PorousFlowFullySaturated adds relevant Kernels
# NOTE  [TensorMechanics/Master]
# NOTE    [all]
# NOTE      add_variables = true
# NOTE      strain = FINITE
# NOTE    []
# NOTE  []
  [FluidProperties]
    [simple_fluid]
      type = SimpleFluidProperties
      bulk_modulus = 1e7 #Pa
      density0 = 1000 #rho in kg/m³
      thermal_expansion = 0
      viscosity = 1e-3 #mu in Pa*s
    []
  []
[]

[PorousFlowFullySaturated]
  coupling_type = HydroMechanical
  porepressure = porepressure
  biot_coefficient = 0.8
  displacements = 'disp_x disp_y'
  fp = simple_fluid
  stabilization = Full
  pressure_unit = MPa
[]

[BCs]
  [confinex]
    type = DirichletBC
    variable = disp_x
    value = 0
    boundary = 'left bottom'
  []
  [confiney]
    type = DirichletBC
    variable = disp_y
    value = 0
    boundary = 'bottom'
  []
  [topload]
    type = FunctionDirichletBC
    variable = disp_y
    function = tdisp_y
    boundary = 110
  []    
  [outflow] # NOTE PorousFlowOutflow might be better here?
    type = PorousFlowPiecewiseLinearSink
    variable = porepressure
    boundary = 111
    pt_vals = '-1e9 1e9'
    multipliers = '-1e9 1e9'
    PT_shift = 0
    flux_function = 1 # NOTE, originally 10.  This is a major problem for convergence
    fluid_phase = 0
    use_mobility = true
    use_relperm = true
  []
[]

[Materials]
  [elasticity_cube]
    type = ComputeIsotropicElasticityTensor
    shear_modulus = 0.5
    bulk_modulus = 320
  []
# NOTE need this because no TensorMechanics Action
# NOTE this is small strain
  [strain]
    type = ComputeSmallStrain
  []
  [stress]
# NOTE change from finite strain
    type = ComputeLinearElasticStress
  []
  [porosity_cube]
    type = PorousFlowPorosity
    fluid = true
    mechanical = true
    porosity_zero = 0.5 
    biot_coefficient = 0.8 
    solid_bulk = 320
  []
  [permeability_cube]
    type = PorousFlowPermeabilityConst
    permeability = '0.25e-9 0 0   0 0.25e-9 0   0 0 0.25e-9'
  []
[]

[Preconditioning]
  [andy]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = 'lu       mumps'
    
    #petsc_options_iname = '-pc_type -pc_hypre_type'
    #petsc_options_value = 'hypre boomeramg'
    
    #petsc_options_iname = '-ksp_type -pc_type -sub_pc_type -sub_pc_factor_shift_type -pc_asm_overlap'
    #petsc_options_value = 'gmres      asm      lu           NONZERO                   2             '
  []
[]

[Executioner]
  type = Transient
  solve_type = Newton
  start_time = 0
  end_time = 10
  dt = 0.5  
  nl_max_its = 100
# NOTE  nl_abs_tol = 1e-3
# NOTE with infinitesimal strain and small flux_function, can choose small nl_rel_tol
  nl_rel_tol = 1e-6
[]

[Outputs]
  csv = true
  exodus = true
[]
```

**Reply by Flolaffel** [reply]
Ok, first of all thanks a lot for the work you put in your answer. I'd like to adress your points one by one, so here goes.

1. **`TimeDerivative` Kernel**.  was a really helpful trick to get my indenter to behave like the massive piece of material it is. 

2. **`TensorMechanicsAction`**. What's the exact reason not to use this action? Because following the documentation the only things I told it to do is add the variables and the strain Kernel. Is there more to it in the background that overlaps with the `FullySaturatedAction`?

4. **Locking correction.** I've noticed the convergence impact `volumetric_locking_correction` has before. See an older thread by me:
#20856
Unluckily there doesn't really seem to be a solution for the negative impact it has on convergence. I've experienced situations in the past where `volumetric_locking_correction` was dearly needed for accurate results when using highly deformable materials, so I just include it everywhere. Though you are right it might not be needed here and could save me some hustle. So thanks for the reminder.

5. **BC.** You are right that `PorousFlowOutflowBC` might be a feasible replacement in the input file I sent you. However in my contact variant I'm in need for a BC that can accept a function input, so `PorousFlowPiecewiseLinearSink` becomes the best option. When setting the `flux_function` value, I used the [PorousFlow BC documentation](https://mooseframework.inl.gov/modules/porous_flow/boundaries.html). There L=10^-4 m is mentioned as an appropiate value to set the sink to when you are aiming at DirichletBC behaviour. That would in my case be 10^-1 mm. With the option `use_mobility = true` the `flux_function` becomes 1/L so 10. Although I'll keep your concern in mind.

Finally, I did some experimenting over the weekend and found the underlying issue that caused the weird behaviour of my porepressure. I'll write that up in a separate answer after lunch.

**Reply by Flolaffel** [reply]
Ok, so: I did some further testing and I think I now reached a satisfactory result.

What caused the weird porepressure behaviour apparently were: the contact and the MasterAction. 

What I discovered while testing contact was that not the contact itself causes a problem but rather the nodes that leave the contact do. I could imagine that a node in contact stores quite some fluid since I don't allow it to get rid of it. And as soon as said node leaves contact, it has to get rid of the fluid which may cause the weird pressure behaviour. Or it's just some numerical trickery I don't understand. Either way, setting contact to `glued` instead of `frictionless` and toning down my deformation solved the problem. Or so I thought.

While doing further testing today, regarding your answer and input, I realized something else: the weird behaviour came back, wenn I reduced my convergence criteria. Everything was fine with `nl_abs_tol = 1e-3`. But when I switched to `nl_abs_tol = 1e-6` combined with `nl_rel_tol = 1e-6` the behaviour was back.

So the second solution actually was disabling the MasterAction. I'm not entirely sure why, see my previous answer. But I replaced it with the needed variables and an `ComputeAxisymmetricRZFiniteStrain` (I think you forgot to set the strain calculation to RZ in your input). I forgot it too at first and was surprised to see different deformation results between using the action and not using it. But then I realized and using the correct strain class yields similar deformation results as the MasterAction but without the porepressure suddenly vanishing.

This is the (for now) final input. Thanks a lot again for your detailed answers. Really insightful.

```
[Mesh]
  [tool]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 5
    ymin = 0
    ymax = 2
    nx = 10
    ny = 5
  []
  [transform]
    type = TransformGenerator
    input = tool
    transform = translate
    vector_value = '0 5 0'
  []
  [tool_id]
    type = SubdomainIDGenerator
    input = transform
    subdomain_id = 1
  []
  [cube]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 10
    ymin = 0
    ymax = 5
    nx = 20
    ny = 15
    bias_y = 0.9
    boundary_id_offset = 100
    boundary_name_prefix = cube
  []
  [cube_id]
    type = SubdomainIDGenerator
    input = cube
    subdomain_id = 2
  []
  [collection]
    type = MeshCollectionGenerator
    inputs = 'tool_id cube_id'
  []
  construct_side_list_from_node_list = true
  patch_update_strategy = iteration
[]

[Problem]
  coord_type = RZ
[]

[GlobalParams]
  displacements = 'disp_x disp_y'
  gravity = '0 0 0'
  PorousFlowDictator = dictator
  #volumetric_locking_correction = true
[]

[Variables]
  [disp_x]
  []
  [disp_y]
  []
  [porepressure]
  []
[]

[AuxVariables]
  [nodal_outflow]
  []
  [porosity]
    family = MONOMIAL
  []
[]

[AuxKernels]
  [porosity]
    type = PorousFlowPropertyAux
    property = porosity
    variable = porosity
    phase = 0
  []
[]

[Functions]
  [tdisp_y]
    type = PiecewiseLinear
    x = '0   100   400'
    y = '0  -0.25 -0.25'
  []
  [vpp]
    type = VectorPostprocessorFunction
    vectorpostprocessor_name = cont_press
    argument_column = x
    value_column = contact_pressure
    component = x
  []
  [outflow]
    type = ParsedFunction
    value = 'if(cp > 0, 0, 10)' # in mm
    vars = 'cp'
    vals = 'vpp'
  []
[]

[Modules]
  [FluidProperties]
    [simple_fluid]
      type = SimpleFluidProperties
      bulk_modulus = 1e7 #Pa, 8
      density0 = 1000 #rho in kg/m³, 1
      thermal_expansion = 0
      viscosity = 1e-3 #mu in Pa*s, 0.96
    []
  []
[]

[PorousFlowFullySaturated]
  coupling_type = HydroMechanical
  porepressure = porepressure
  biot_coefficient = 0.8 #0.6
  displacements = 'disp_x disp_y'
  fp = simple_fluid
  stabilization = Full
  pressure_unit = MPa
[]

[Contact]
  [kontakt]
    primary = bottom
    secondary = cube_top
    model = glued
    formulation = penalty
    normalize_penalty = true
    penalty = 1e4
    tangential_tolerance = 0.005
  []
[]

[BCs]
  [confinex]
    type = DirichletBC
    variable = disp_x
    value = 0
    boundary = 'cube_left top cube_bottom'
  []
  [confiney]
    type = DirichletBC
    variable = disp_y
    value = 0
    boundary = 'cube_bottom'
  []
  [topload]
    type = FunctionDirichletBC
    variable = disp_y
    function = tdisp_y
    boundary = top
  []  
  [outflow]
    type = PorousFlowPiecewiseLinearSink
    variable = porepressure
    boundary = cube_top
    pt_vals = '0 1e9'
    multipliers = '0 1e9'
    PT_shift = 0
    flux_function = outflow 
    fluid_phase = 0
    use_mobility = true
  []
[]

[Kernels]
  [dummy_p]
    type = TimeDerivative
    variable = porepressure
    block = 1
  []
[]

[Materials]
  [strain]
    type = ComputeAxisymmetricRZFiniteStrain
    block = '1 2'
  []
  [elasticity_tool]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 1.0e6
    poissons_ratio = 0.25
    block = 1
  []
  [elasticity_cube]
    type = ComputeIsotropicElasticityTensor
    bulk_modulus = 320
    shear_modulus = 0.5
    block = 2
  []
  [stress_tool]
    type = ComputeFiniteStrainElasticStress
    block = '1 2'
  []
  [porosity_tool]
    type = PorousFlowPorosityConst
    porosity = 0
    block = 1
  []
  [porosity_cube]
    type = PorousFlowPorosity
    porosity_zero = 0.5
    biot_coefficient = 0.8 
    solid_bulk = 320 
    fluid = true
    mechanical = true
    block = 2
  []
  [permeability_tool]
    type = PorousFlowPermeabilityConst
    permeability = '0 0 0   0 0 0   0 0 0'
    block = 1
  []
  [permeability_cube]
    type = PorousFlowPermeabilityConst
    permeability = '0.25e-9 0 0   0 0.25e-9 0   0 0 0.25e-9' #k in mm^2
    block = 2
  []
[]

[Postprocessors]
  [mass_ph0]
    type = PorousFlowFluidMass
    fluid_component = 0
    execute_on = 'initial timestep_end'
    block = 2
  []
[]

[VectorPostprocessors]
  [cont_press]
    type = NodalValueSampler
    variable = contact_pressure
    boundary = cube_top
    sort_by = x
  []
  [print_vpp]
    type = LineFunctionSampler
    functions = vpp
    start_point = '0 5 0'
    end_point = '10 5 0'
    num_points = 21
    sort_by = id
  []
  [print_outflow]
    type = LineFunctionSampler
    functions = outflow
    start_point = '0 5 0'
    end_point = '10 5 0'
    num_points = 21
    sort_by = id
  []
[]

[Preconditioning]
  [andy]
    type = SMP
    full = true
    petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
    petsc_options_value = 'lu       mumps'
  []
[]

[Executioner]
  type = Transient
  solve_type = Newton
  start_time = 0
  end_time = 400
  dt = 5
  line_search = basic
  nl_abs_tol = 1e-6
  nl_rel_tol = 1e-6
[]

[Outputs]
  csv = true
  exodus = true
[]
```

**Reply by WilkAndy** [reply]
I'm so glad that you tracked the problem down, phew!  It's great you're happy with the solution.

For future readers: The reason you don't want the `TensorMechanics Action` is that `PorousFlowFullySaturated` already adds the PDEs for solid poro-mechanics (that is, solid mechanics with effective stress).  As i understand it, the `TM Action` **does** add Kernels.  Hence, if you include the `TM Action`, you'll be solving a strange summation of the usual solid mechanics (provided by TM - note that without auto-differentiation, the Jacobians are known to be non-exact so you might get poor convergence) along with poromechanics (provided by `FullySaturated`).  

For future readers: ooops, yes, the `RZ` strain is the correct one to use here!
